transport/transport.go

```diff
@@ -0,0 +1,113 @@
+package transport
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"math/big"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Transport interface {
+	// NamespacedName returns the namespaced name to identify this transport Transport
+	NamespacedName() types.NamespacedName
+	// ListenPort returns a port on which the transport server listens for incomming connections
+	ListenPort() int32
+	// ConnectPort returns the port to connect to transfer server
+	// Using this the server acts as a client to the transfer relaying all the data
+	// sent from the transport client
+	ConnectPort() int32
+	// Containers returns a list of containers transfers can add to their server Pods
+	Containers() []corev1.Container
+	// Volumes returns a list of volumes transfers have add to their server Pods for getting the configurations
+	// mounted for the transport containers to work
+	Volumes() []corev1.Volume
+	// Type
+	Type() Type
+	// Credentials returns the namespaced name of the secret holding credentials for talking to the server
+	Credentials() types.NamespacedName
+	// Hostname returns the string to which the transfer will connect to
+	// in case of a null transport, it will simple relay the endpoint hostname
+	// in case of a valid transport, it will have a custom hostname where transfers will have to connect to.
+	Hostname() string
+	// MarkForCleanup adds a label to all the resources created for the endpoint
+	// Callers are expected to not overwrite
+	MarkForCleanup(ctx context.Context, c client.Client, key, value string) error
+}
+
+type Options struct {
+	Labels map[string]string
+	Owners []metav1.OwnerReference
```

Comment:
I notice when this gets used, it is directly assigned to the object's field instead of going through something like https://github.com/kubernetes-sigs/controller-runtime/blob/5f8befe79888a4a8529cda7a82797bd57e188158/pkg/controller/controllerutil/controllerutil.go#L96 that validates the reference (i.e., owner scoping, same namespace, etc). Does this get validated elsewhere?

Reply:
https://github.com/backube/pvc-transfer/issues/10